### PR TITLE
Update with a test data of the staggered scheme

### PR DESCRIPTION
**Pull request by wenqing**

### CODE REVIEW

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 1]
```diff
@@ -0,0 +1,304 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<OpenGeoSysProject>
+    <mesh>mesh2D.vtu</mesh>
+    <geometry>gravity_driven.gml</geometry>
+
+    <processes>
+        <process>
+            <name>HeatConduction</name>
+            <type>HEAT_CONDUCTION</type>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:9]: A general comment: there needs to be some kind of naming convention for processes, parameters etc. Here, the processes type has capital letters and underscores, in the case of monolothic TH, it is camel case (ConstViscosityThermalConvection)

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            <secondary_variables>
+                <secondary_variable type="static" internal_name="heat_flux_x" output_name="heat_flux_x" />
+            </secondary_variables>
+        </process>
+        <process>
+            <name>LiquidFlow</name>
+            <type>LIQUID_FLOW</type>
+            <integration_order>2</integration_order>
+
+            <darcy_gravity>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:29]: Similar to above: here there needs to be some consistency. LIQUID_FLOW wants ```darcy_gravity```, TH monolithic takes ```specific_body_force```

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+               <!-- axis_id: 0, 1, or the dimension of space minus one -->
+               <axis_id>1</axis_id>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:31]: what if you want to have a gravity vector in a direction that is not parallel to any axis of the coordinate system, ie for y and z?
I would prefer the writing style as in TH mono:
```<specific_body_force>0 9.81</specific_body_force>```

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+    <processes>
+        <process>
+            <name>HeatConduction</name>
+            <type>HEAT_CONDUCTION</type>
+            <integration_order>2</integration_order>
+
+            <thermal_conductivity>lambda</thermal_conductivity>
+            <heat_capacity>c_p</heat_capacity>
+            <density>rho</density>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:14]: why do I have to give ```rho``` here, if in L47 I give it again?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+            <secondary_variables>
+                <secondary_variable type="static" internal_name="darcy_velocity_x" output_name="v_x" />
+                <secondary_variable type="static" internal_name="darcy_velocity_y" output_name="v_y" />
+            </secondary_variables>
+
+            <material_property>
+                <fluid>
+                    <density>
+                       <type>TemperatureDependent</type>
+                       <temperature0> 293.0 </temperature0>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:49]: are spaces important?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+       <global_process_coupling>
+           <max_iter> 5 </max_iter>
+           <convergence_criterion>
+               <type>Residual</type>
+               <norm_type>NORM2</norm_type>
+               <reltol>1.e-4</reltol>
+           </convergence_criterion>
+       </global_process_coupling>
+        <processes>
+            <process ref="HeatConduction">
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:99]: Could I also turn this around and give LiquidFlow here, and down below at the coupled_process HeatConduction?

Comment (hydromarc) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:99]: Oh, I see, that this is done anyway again... so how does the coupling loop work? Is it firstly running heat then liquid, and as a second step liquid then heat?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+            <type>Constant</type>
+            <value>3.0</value>
+        </parameter>
+        <parameter>
+            <name>c_p</name>
+            <type>Constant</type>
+            <value>1000</value>
+        </parameter>
+        <parameter>
+            <name>rho</name>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:184]: why is this needed here?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+                            <value> 0.2 </value>
+                        </porosity>
+                        <storage>
+                            <type>Constant</type>
+                            <value> 0.0 </value>
+                        </storage>
+                    </porous_medium>
+                </porous_medium>
+                <solid>
+                    <thermal_expansion>solid_thermal_expansion</thermal_expansion>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:82]: will this be included into the fluid/material properties once?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+            <name>basic_picard_H</name>
+            <type>Picard</type>
+            <max_iter>10</max_iter>
+            <linear_solver>linear_solver_H</linear_solver>
+        </nonlinear_solver>
+    </nonlinear_solvers>
+
+    <linear_solvers>
+        <linear_solver>
+            <name>linear_solver_T</name>
```
Comment (hydromarc) [inline StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:275]: where or through which tag do I assign the respective processes to the two listed solvers here?

**Review by hydromarc — COMMENTED** [review]
@wenqing Could you please reply to my comments? Thanks!

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 1 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:9]: It is defined in the source code for process type. See: https://github.com/ufz/ogs/blob/master/Applications/ApplicationsLib/ProjectData.cpp#L303 (from line 303).

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 4 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:14]: It is a Parameter data name, and its value is given in the segment of <parameters>.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 2 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:29]: This could be discussed.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 3 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:31]: I think this a case of inclined 1D in 2D/3D space or 2D element in 3D space. A transform of vector and integration Jacobian are needed also for the style in TH mono.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 8 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:82]: No, this is for solid and it takes the Parameter type data, which is a sort of global data.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 6 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:99]: Just follows the order of the processes defined in this project file. For instance, if heat process is defined first, then the coupling loop solves the heat process  equation firstly.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 7 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:184]: Here the value of rho is defined.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 9 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:275]: By the order of definitions of processes. The first one for the first process, the second for the second process, etc.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 5 above)
Comment (wenqing) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:49]: No.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 1 above)
Comment (hydromarc) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:9]: Ah, yes - good, my mistake.

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 4 above)
Comment (hydromarc) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:14]: Ok, but "rho" is the fluid density here?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 2 above)
Comment (hydromarc) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:29]: yes, please do this (probably @endJunction just start and supervise this discussion)

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 3 above)
Comment (hydromarc) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:31]: ok, will you adjust this?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 8 above)
Comment (hydromarc) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:82]: Yet, this could be a parameter that is material-specific, right?

File: StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj (hunk 6 above)
Comment (hydromarc) [reply StaggeredCoupledProcesses/TH/ThermalDensityDrivenFlow2D/thermal_gravity_driven_flow2d.prj:99]: Good.
